alphagov/whitehall
# Add GA4 tracking to HTML attachment links

Hi @andysellick / @JamesCGDS - would you be able to approve this if all is OK?

## What

- Adds GA4 link tracking to HTML attachment links

## Why

- It's part of our interactions for our GA4 tracking.

Trello card: https://trello.com/c/nJ05htIn/491-add-tracking-attachment-link-clicks

## Testing

It's a bit difficult to Test as getting a local Whitehall page up isn't too easy at the moment. I managed to render the attachment HTML component on the Whitehall Admin dashboard instead (thanks @andysellick for the idea). Then I checked that the data attributes were correctly attached to the HTML from there. If you want to see the HTML rendered, what I did was:

- Run `make whitehall` then `govuk-docker up whitehall-app`
- In `whitehall`, copy everything from `app/views/documents/_attachment.html.erb` and paste it into `app/views/admin/dashboard/index.html.erb`
- Add `attachment = ExternalAttachment.new()` above the attachment code in `index.html.erb`
- Go to `app/models/external_attachment.rb` and find the method `def html?`. Change it so that it returns `true`. 
- If you go to http://whitehall-admin.dev.gov.uk/government/admin/ you'll see the HTML attachment with GA4 attributes on it. The CSS will be broken but that isn't needed here.
- There's no GA4 JavaScript in the Whitehall Admin dashboard so actually clicking the link won't do anything as there's no `dataLayer` or GA4 modules. But what you can do is copy the `ga4-link` JSON, and paste it onto a link on one of our GOVUK homepage links. Then, if you clicked the link you'll see that the attachment link JSON I wrote works fine with our link tracker. 

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.
